### PR TITLE
Activating holo hosted happs

### PR DESCRIPTION
**Pull request by zo-el**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -32,6 +33,11 @@ git = "https://github.com/holochain/holochain"
 rev = "ea026def01d1573d5e0edfb7f4e3e9453f88c43e"
 package = "holochain_types"
 
+[dependencies.holochain_types]
```
Comment (Br1ght0ne) [inline Cargo.toml:36]: Would be best to have the dependency name correspond.

```suggestion
[dependencies.holochain_zome_types]
```

**Review by Br1ght0ne — COMMENTED** [review]
I also have some naming suggestions, but I'd leave that until I continue working on this PR. Good work @zo-el!

File: Cargo.toml (hunk 1 above)
Comment (Br1ght0ne) [reply Cargo.toml:36]: This is actually required to build correctly.

File: src/lib.rs [hunk 2]
```diff
@@ -17,8 +20,95 @@ use tempfile::TempDir;
...
+                zome_name: ZomeName::from("hha"),
+                fn_name: FunctionName::from("get_happs"),
+                payload: ExternInput::new(SerializedBytes::default()),
+                cap: None,
+                provenance: app_info.cell_data[0]
+                    .clone()
+                    .into_id()
+                    .agent_pubkey()
+                    .to_owned(),
+            };
```
Comment (timotree3) [inline src/lib.rs:88]: this business with cloning provenance two times looks ugly/inefficient. Here's a way of rewriting this to take ownership of the necessary values and avoid cloning
```suggestion
        Some(holochain_types::InstalledApp {
            // This works on the assumption that the core happs has HHA in the first position of the vec
            cell_data: [hha_cell, ..],
            ..
        }) => {
            let zome_call_payload = ZomeCall {
                cell_id: hha_cell.as_id().clone(),
                zome_name: ZomeName::from("hha"),
                fn_name: FunctionName::from("get_happs"),
                payload: ExternInput::new(SerializedBytes::default()),
                cap: None,
                provenance: hha_cell
                    .into_id()
                    .into_dna_and_agent()
                    .1
            };
```

**Review by timotree3 — APPROVED** [review]
Good work :) I have one suggestion about unnecessary cloning but the functionality looks great to me

File: src/lib.rs [hunk 3]
```diff
@@ -17,8 +20,95 @@ use tempfile::TempDir;
...
+            .json(&body)
+            .send()
+            .await?;
+        info!("Installed happ-id {:?}", happ_id);
+        info!("Response {:?}", response);
+    }
+    Ok(())
+}
+
+pub async fn get_enabled_hosted_happs(core_happ: Happ) -> Result<Vec<WrappedHeaderHash>> {
```
Comment (timotree3) [inline src/lib.rs:71]: I have an allergic reaction to allocations when coding in Rust. Here we're returning a `Vec` (requires allocation) by calling `.collect` on an iterator and all we're doing with it at the call site is iterating over it. It would avoid that allocation and therefore be marginally more efficient to return an `impl Iterator` instead

That requires changing three things: The return type of this function to
```suggestion
pub async fn get_enabled_hosted_happs(core_happ: Happ) -> Result<impl Iterator<Item=WrapperHeaderHash>> {
```

The return value of this function
```diff
-                     let happ_bundle_ids: Vec<WrappedHeaderHash> =
-                         .collect();
-                     info!("List of happ_ids {:?}", happ_bundle_ids);
+                    let happ_bundle_ids = happ_bundles.into_iter().map(|happ| happ.happ_id);
```

and the argument type of the function that iterates over it

```diff
- pub async fn install_holo_hosted_happs(happs: Vec<WrappedHeaderHash>) -> Result<()> {
+ pub async fn install_holo_hosted_happs(happs: impl Iterator<Item=WrappedHeaderHash>) -> Result<()> {
```

File: src/lib.rs [hunk 4]
```diff
@@ -17,8 +20,95 @@ use tempfile::TempDir;
...
+                        happ_bundles.into_iter().map(|happ| happ.happ_id).collect();
+                    info!("List of happ_ids {:?}", happ_bundle_ids);
+                    return Ok(happ_bundle_ids);
+                }
+                _ => return Err(anyhow!("unexpected response: {:?}", response)),
+            }
+        }
+        None => {
+            return Err(anyhow!("HHA is not installed"));
+        }
```
Comment (timotree3) [inline src/lib.rs:108]: No need to use `return` at the end of a scope
```suggestion
                    Ok(happ_bundle_ids)
                }
                _ => Err(anyhow!("unexpected response: {:?}", response)),
            }
        }
        None => Err(anyhow!("HHA is not installed")),
```

**Review by timotree3 — COMMENTED** [review]
Now that you asked me to give a more critical review I'll add a couple more things that I didn't pay much attention to in the first pass

**Comment by peeech** [comment]
This is such a great review! Thanks for teaching us something new!!!

File: src/lib.rs (hunk 3 above)
Comment (zo-el) [reply src/lib.rs:71]: The only reason I think I have preferred using `Vec` is because it is easier to debug. It looks like `Debug` is not implemented for `impl Iterator`

File: src/lib.rs (hunk 2 above)
Comment (zo-el) [reply src/lib.rs]: It looks like `Vec` cannot be de-structured like `[happ_cell, ...]`.

File: src/lib.rs (hunk 3 above)
Comment (timotree3) [reply src/lib.rs]: Returning `impl Trait` means that you're returning some specific type that implements that trait, and you just don't want to say which type. The actual type here in `std::iter::Map<std::vec::IntoIter<WrapperHeaderHash>, F>` where `F` is the type of the closure `|happ| happ.happ_id`

That type does implement `Debug`, so in a context where you have the full type such as within this function, it's already possible to print out the iterator. In another context where the "type" is `impl Iterator`, such as `install_holo_hosted_happs`, one solution would be to add a `Debug` bound: replace `impl Iterator<Item=WrappedHeaderHash>` with `impl Iterator<Item=WrapperHeaderHash> + Debug`

Note that when you debug one of these iterators it would probably only print out the contents that have not yet been consumed.

File: src/lib.rs (hunk 2 above)
Comment (timotree3) [reply src/lib.rs]: Darn it!! I'll make a new suggestion using `into_iter()` to take ownership then

File: src/lib.rs [hunk 5]
```diff
@@ -17,8 +20,94 @@ use tempfile::TempDir;
...
+            ..
+        }) => {
+            let zome_call_payload = ZomeCall {
+                cell_id: cell_data[0].as_id().clone(),
+                zome_name: ZomeName::from("hha"),
+                fn_name: FunctionName::from("get_happs"),
+                payload: ExternInput::new(SerializedBytes::default()),
+                cap: None,
+                provenance: cell_data[0].clone().into_id().into_dna_and_agent().1,
+            };
```
Comment (timotree3) [inline src/lib.rs:91]: Let me try again at improving this cloning.
```suggestion
        Some(holochain_types::prelude::InstalledApp {
            cell_data,
            ..
        }) => {
            let hha_cell = cell_data.into_iter().next().unwrap_or_else(|| anyhow!("expected HHA to have at least one cell")?;
            let provenance = hha_cell.as_id().agent_pubkey().clone();
            let zome_call_payload = ZomeCall {
                cell_id: hha_cell.into_id(),
                zome_name: ZomeName::from("hha"),
                fn_name: FunctionName::from("get_happs"),
                payload: ExternInput::new(SerializedBytes::default()),
                cap: None,
                provenance,
            };
```
